Spigot-Server-Patches/0236-Anti-Xray.patch

```diff
@@ -0,0 +1,1560 @@
+             this.a(packet, (GenericFutureListener[]) null);
+         } else {
+             this.j.writeLock().lock();
+@@ -154,8 +154,8 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+     }
+ 
+     public void sendPacket(Packet<?> packet, GenericFutureListener<? extends Future<? super Void>> genericfuturelistener, GenericFutureListener<? extends Future<? super Void>>... agenericfuturelistener) {
+-        if (this.isConnected()) {
+-            this.m();
++        if (this.isConnected() && this.trySendQueue() && !(packet instanceof PacketPlayOutMapChunk && !((PacketPlayOutMapChunk) packet).isReady())) { // Paper - Async-Anti-Xray - Add chunk packets which are not ready or all packets if the queue contains chunk packets which are not ready to the queue and send the packets later in the right order
++            //this.m(); // Paper - Async-Anti-Xray - Move to if statement (this.trySendQueue())
+             this.a(packet, (GenericFutureListener[]) ArrayUtils.add(agenericfuturelistener, 0, genericfuturelistener));
+         } else {
+             this.j.writeLock().lock();
+@@ -169,6 +169,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+ 
+     }
+ 
++    private void dispatchPacket(final Packet<?> packet, @Nullable final GenericFutureListener<? extends Future<? super Void>>[] genericFutureListeners) { this.a(packet, genericFutureListeners); } // Paper - Anti-Xray - OBFHELPER
+     private void a(final Packet<?> packet, @Nullable final GenericFutureListener<? extends Future<? super Void>>[] agenericfuturelistener) {
+         final EnumProtocol enumprotocol = EnumProtocol.a(packet);
+         final EnumProtocol enumprotocol1 = (EnumProtocol) this.channel.attr(NetworkManager.c).get();
+@@ -210,22 +211,36 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet<?>> {
+ 
+     }
+ 
+-    private void m() {
++    // Paper start - Async-Anti-Xray - Stop dispatching further packets and return false if the peeked packet is a chunk packet which is not ready
++    private boolean trySendQueue() { return this.m(); } // OBFHELPER
++    private boolean m() { // void -> boolean
+         if (this.channel != null && this.channel.isOpen()) {
+             this.j.readLock().lock();
+ 
+             try {
+                 while (!this.i.isEmpty()) {
+-                    NetworkManager.QueuedPacket networkmanager_queuedpacket = (NetworkManager.QueuedPacket) this.i.poll();
+-
+-                    this.a(networkmanager_queuedpacket.a, networkmanager_queuedpacket.b);
++                    NetworkManager.QueuedPacket networkmanager_queuedpacket = (NetworkManager.QueuedPacket) this.getPacketQueue().peek(); // poll -> peek
++
++                    if (networkmanager_queuedpacket == null) { // Fix NPE (possible vanilla bug caused by handleDisconnection())
++                        this.getPacketQueue().poll(); // The queue is probably empty anyway
```

Comment:
This isn't really safe.  It may of been empty, but then before .poll is called, it became non empty.

prob best to wrap this in a !isEmpty()

Reply:
Sorry I don't get you. How should it suddenly become non empty? Did you realize the lock?

Reply:
I missed the lock, though I question with proper usage of a CLQ, do we even need the lock?

Reply:
The lock is vanilla implementation. I think the lock is not necessary in the vanilla code but the Anti-Xray implementation makes use of it.